lib/elastic_apm/transport/base.rb

```diff
@@ -8,90 +8,152 @@
 
 module ElasticAPM
   module Transport
+    # rubocop:disable Metrics/ClassLength
     # @api private
     class Base
       include Logging
 
+      WATCHER_EXECUTION_INTERVAL = 5
+      WATCHER_TIMEOUT_INTERVAL = 4
+      WORKER_JOIN_TIMEOUT = 5
+
       def initialize(config)
         @config = config
         @queue = SizedQueue.new(config.api_buffer_size)
-        @pool = Concurrent::FixedThreadPool.new(config.pool_size)
-        @workers = []
 
         @serializers = Serializers.new(config)
         @filters = Filters.new(config)
+
+        @stopped = Concurrent::AtomicBoolean.new
+        @mutex = Mutex.new
+        @workers = Array.new(config.pool_size)
       end
 
-      attr_reader :config, :queue, :workers, :filters
+      attr_reader :config, :queue, :filters, :workers, :watcher, :stopped
 
       def start
+        debug '%s: Starting Transport', pid_str
+
+        ensure_watcher_running
+        ensure_worker_count
       end
 
       def stop
+        debug '%s: Stopping Transport', pid_str
+
+        @stopped.make_true
+
+        stop_watcher
         stop_workers
       end
 
+      # rubocop:disable Metrics/MethodLength, Metrics/LineLength
       def submit(resource)
+        if @stopped.true?
+          warn '%s: Transport stopping, no new events accepted', pid_str
+          return false
+        end
+
+        ensure_watcher_running
         queue.push(resource, true)
 
-        ensure_worker_count
+        true
       rescue ThreadError
-        warn 'Queue is full (%i items), skipping…', config.api_buffer_size
+        warn '%s: Queue is full (%i items), skipping…', pid_str, config.api_buffer_size
+        nil
+      rescue Exception => e
+        error '%s: Failed adding to the transport queue: %p', pid_str, e.inspect
         nil
       end
+      # rubocop:enable Metrics/MethodLength, Metrics/LineLength
 
       def add_filter(key, callback)
         @filters.add(key, callback)
       end
 
       private
 
+      def pid_str
+        format('[PID:%s]', Process.pid)
+      end
+
+      def ensure_watcher_running
+        # pid has changed == we've forked
+        return if @pid == Process.pid
+
+        @mutex.synchronize do
+          @pid = Process.pid
```

Comment:
You should do another `@pid == Process.pid` before setting the pid. That would prevent two different threads triggering the `Concurrent::TimerTask`. (They might both access the synchronized code, but the second one would return after the check, as now the pid is already the current one).